### PR TITLE
[IoTDB-1989] IoTDB support insert data from Spark

### PR DESCRIPTION
**Pull request by xuanronaldo**
See IOTDB-1989.

### CODE REVIEW

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/tools/DataFrameTools.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.*;
+import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.DOUBLE;
```
Comment (HTHou) [inline spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/tools/DataFrameTools.java:37]: Avoiding using `*` is better.

**Comment by xuanronaldo** [comment]
Ok，got this




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2021年11月29日(星期一) 上午10:58
收件人: ***@***.***&gt;; 
抄送: "Xuan ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/iotdb] [IoTDB-1989] IoTDB support insert data from Spark (PR #4477)





 
@HTHou commented on this pull request.
 
 
In spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/tools/DataFrameTools.java:
 &gt; + +import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.*; +import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.DOUBLE;  
Avoiding using * is better.
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/DefaultSource.scala [hunk 2]
```diff
@@ -34,10 +37,27 @@ private[iotdb] class DefaultSource extends RelationProvider with DataSourceRegis
 
     val iotdbOptions = new IoTDBOptions(parameters)
 
-    if (iotdbOptions.url == null || iotdbOptions.sql == null) {
-      sys.error("IoTDB url or sql not specified")
+    if (iotdbOptions.sql.equals("")) {
+      sys.error("sql not specified")
     }
```
Comment (HTHou) [inline spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/DefaultSource.scala:42]:
```suggestion
    if ("".equals(iotdbOptions.sql)) {
      sys.error("sql not specified")
    }
```

**Comment by xuanronaldo** [comment]
ok，got this




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2021年11月29日(星期一) 中午11:07
收件人: ***@***.***&gt;; 
抄送: "Xuan ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/iotdb] [IoTDB-1989] IoTDB support insert data from Spark (PR #4477)





 
@HTHou commented on this pull request.
 
 
In spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/DefaultSource.scala:
 &gt; +    if (iotdbOptions.sql.equals("")) { +      sys.error("sql not specified")      }  ⬇️ Suggested change -    if (iotdbOptions.sql.equals("")) { -      sys.error("sql not specified") -    } +    if ("".equals(iotdbOptions.sql)) { +      sys.error("sql not specified") +    }  
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: spark-iotdb-connector/pom.xml [hunk 3]
```diff
@@ -97,22 +74,31 @@
...
             <groupId>org.scalatest</groupId>
             <artifactId>scalatest_2.11</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>net.jpountz.lz4</groupId>
+            <artifactId>lz4</artifactId>
+            <version>1.3.0</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (HTHou) [inline spark-iotdb-connector/pom.xml:101]: lz4 in master branch has upgraded to 1.8.0, better to keep the same version.

BTW, what this dependency use for here?

**Comment by xuanronaldo** [comment]
It is confilcted with the Spark, so it is marked by `provided`.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2021年11月29日(星期一) 中午11:13
收件人: ***@***.***&gt;; 
抄送: "Xuan ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/iotdb] [IoTDB-1989] IoTDB support insert data from Spark (PR #4477)





 
@HTHou commented on this pull request.
 
 
In spark-iotdb-connector/pom.xml:
 &gt; +        <dependency&gt; +            <groupId&gt;net.jpountz.lz4</groupId&gt; +            <artifactId&gt;lz4</artifactId&gt; +            <version&gt;1.3.0</version&gt; +            <scope&gt;provided</scope&gt; +        </dependency&gt;  
lz4 in master branch has upgraded to 1.8.0, better to keep the same version.
 
BTW, what this dependency use for here?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: spark-iotdb-connector/pom.xml (hunk 3 above)
Comment (xuanronaldo) [reply spark-iotdb-connector/pom.xml:101]: It is confilcted with the Spark, so it is marked by `provided`.

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/DefaultSource.scala (hunk 2 above)
Comment (xuanronaldo) [reply spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/DefaultSource.scala:42]: ok，got this

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/tools/DataFrameTools.java (hunk 1 above)
Comment (xuanronaldo) [reply spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/tools/DataFrameTools.java:37]: ok，got this

**Comment by HTHou** [comment]
Hi, would you like to add the guide to tell users how to use it?

**Comment by xuanronaldo** [comment]
Yep, I got some report need to do now. I'll add some ITs and guide these days.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2021年11月29日(星期一) 下午4:54
收件人: ***@***.***&gt;; 
抄送: "Xuan ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/iotdb] [IoTDB-1989] IoTDB support insert data from Spark (PR #4477)





 
Hi, would you like to add the guide to tell users how to use it?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by xuanronaldo** [comment]
> Hi, would you like to add the guide to tell users how to use it?

Yep, I got some report need to do now. I'll add some ITs and guide these days.

File: spark-iotdb-connector/pom.xml (hunk 3 above)
Comment (xuanronaldo) [reply spark-iotdb-connector/pom.xml:101]: > 

Yep, it throwed an error `java.lang.NoSuchMethodError: net.jpountz.lz4.LZ4BlockInputStream.<init>`.

Comment (xuanronaldo) [reply spark-iotdb-connector/pom.xml:101]: > 

I merged the code, and delete the dependency of the spark pom. It works now.

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/Converter.scala [hunk 4]
```diff
@@ -65,7 +65,11 @@ object Converter {
       }
 
       val colCount = resultSetMetaData.getColumnCount
-      for (i <- 2 to colCount) {
+      var startIndex = 2
+      if (options.sql.toLowerCase.contains("count")) {
```
Comment (SilverNarcissus) [inline spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/Converter.scala:69]: All aggregation function should be filtered. A better way is check whether the first index column is "time".

File: spark-iotdb-connector/src/test/scala/org/apache/iotdb/spark/db/IoTDBTest.scala [hunk 5]
```diff
@@ -20,7 +20,6 @@
 package org.apache.iotdb.spark.db
 
 import java.io.ByteArrayOutputStream
-
```
Comment (SilverNarcissus) [inline spark-iotdb-connector/src/test/scala/org/apache/iotdb/spark/db/IoTDBTest.scala:23]: We should add some insert test and check whether the result is correct.

**Review by SilverNarcissus — CHANGES_REQUESTED** [review]
Fantastic work! There are 2 issues and failed CI, please have a look.

File: spark-iotdb-connector/src/test/scala/org/apache/iotdb/spark/db/IoTDBTest.scala (hunk 5 above)
Comment (xuanronaldo) [reply spark-iotdb-connector/src/test/scala/org/apache/iotdb/spark/db/IoTDBTest.scala:23]: Yep, I have updated some ITs of insert test.

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/Converter.scala (hunk 4 above)
Comment (xuanronaldo) [reply spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/Converter.scala:69]: Ok, got this.

**Review by SilverNarcissus — APPROVED** [review]
Look good to me~

File: docs/zh/UserGuide/Ecosystem Integration/Spark IoTDB.md [hunk 6]
```diff
@@ -43,7 +43,9 @@ mvn clean scala:compile compile install
     </dependency>
 ```
 
-#### Spark-shell 用户指南
+## 从IoTDB导入数据
```
Comment (HTHou) [inline docs/zh/UserGuide/Ecosystem Integration/Spark IoTDB.md:46]: 从iotdb 读取数据

Comment (xuanronaldo) [reply docs/zh/UserGuide/Ecosystem Integration/Spark IoTDB.md:46]: got it

**Review by HTHou — APPROVED** [review]

File: docs/UserGuide/Ecosystem Integration/Spark IoTDB.md [hunk 7]
```diff
@@ -158,3 +159,29 @@ public class Example {
...
+  .withColumnRenamed("_8", "text")
+dfWithColumn
+    .write
+    .format("org.apache.iotdb.spark.db")
+    .option("url", "jdbc:iotdb://127.0.0.1:6667/")
+    .save
+```
+
+### Notes
+1. You can directly write data to IoTDB whatever the dataframe contains a wide table or a narrow table.
```
Comment (HTHou) [inline docs/UserGuide/Ecosystem Integration/Spark IoTDB.md:187]: Give an example of importing wide table?

**Review by HTHou — CHANGES_REQUESTED** [review]
Plz look the comments above.

File: docs/UserGuide/Ecosystem Integration/Spark IoTDB.md (hunk 7 above)
Comment (xuanronaldo) [reply docs/UserGuide/Ecosystem Integration/Spark IoTDB.md:187]: got this

**Review by HTHou — APPROVED** [review]

**Pull request merged**